README.md
# TextTable

[![Swift 3.0](https://img.shields.io/badge/Swift-3.0-orange.svg?style=flat)](https://swift.org)
[![License MIT](https://img.shields.io/badge/License-MIT-blue.svg?style=flat)](https://tldrlegal.com/license/mit-license)
[![License MIT](https://img.shields.io/badge/SPM-1.0.0--alpha.4-red.svg?style=flat)](https://github.com/cfilipov/TextTable/releases/tag/v1.0.0-alpha.4)
[![Build Status](https://travis-ci.org/cfilipov/TextTable.svg?branch=master)](https://travis-ci.org/cfilipov/TextTable)

Easily print textual tables in Swift. Inspired by the Python [tabulate](https://pypi.python.org/pypi/tabulate) library.

## Features

* Easily generate textual tables from collections of any type
* Auto-calculates column width
* Custom width for individual columns
* Align individual columns
* Supports per-column [Formatters](https://developer.apple.com/reference/foundation/nsformatter)
* Multiple table output styles including Markdown, Emacs, HTML, Latex etc...
* Extensible table styles allow you to easily create your own table style
* Column-specific truncation modes (truncate head or tail of column content or optionally fail if content overflows).

## Requirements

This package was written for and tested with the following version of Swift:

> Apple Swift version 3.0 (swiftlang-800.0.43.6 clang-800.0.38)

## Usage

Use the [Swift Package Manager](https://swift.org/package-manager/) to install `TextTable` as a dependency of your project.

```Swift
dependencies: [
    .Package(
        url: "https://github.com/cfilipov/TextTable",
        Version(1, 0, 0, prereleaseIdentifiers: ["alpha", "4"]))
]
```

First, import `TextTable`:

```Swift
import TextTable
```

Throughout the examples, the following data structure will be used:

```Swift
struct Person {
    let name: String
    let age: Int
    let birhtday: Date
}
```

Let's say we have a collection of `Person` we wish to display as a table.

```Swift
let data = [
	Person(name: "Alice", age: 42, birhtday: Date()),
	Person(name: "Bob", age: 22, birhtday: Date()),
	Person(name: "Eve", age: 142, birhtday: Date())
]
```

### Configure a TextTable Instance

Configure a `TextTable` instance for the type you wish to print as a table. The minimum configuration will require a mapping of values to columns for your type.

```Swift
let table = TextTable<Person> {
    [Column("Name" <- $0.name),
     Column("Age" <- $0.age),
     Column("Birthday" <- $0.birhtday)]
}
```

#### Alternate Syntax

The `<-` operator is just syntactic sugar. If you don't want to use the operator, you can use labeled arguments instead.

```Swift
let table = TextTable<Person> {
    [Column(title: "Name", value: $0.name),
     Column(title: "Age", value: $0.age),
     Column(title: "Birthday", value: $0.birhtday)]
}
```

### Printing a Table

Simply call the `print()` method, passing in a collection of `Person`.

```Swift
table.print(data)
```
```
Name  Age Birthday
----- --- -------------------------
Alice 42  2016-08-13 19:21:54 +0000
Bob   22  2016-08-13 19:21:54 +0000
Eve   142 2016-08-13 19:21:54 +0000
```

### Getting a String

You can also just get the rendered table as a string by calling the `string(for:)` method.

```Swift
let s = table.string(for: data)
```

### Customize Table Style

You can optionally specify a table style using the `style:` argument.

```Swift
table.print(data, style: Style.psql)
```
```
+-------+-----+---------------------------+
| Name  | Age | Birthday                  |
+-------+-----+---------------------------+
| Alice | 42  | 2016-08-13 08:12:58 +0000 |
| Bob   | 22  | 2016-08-13 08:12:58 +0000 |
| Eve   | 142 | 2016-08-13 08:12:58 +0000 |
+-------+-----+---------------------------+
```

The style can also be specified when using the `string(for:)` method.

```Swift
let s = table.string(for: data, style: Style.psql)
```

For a full list of supported styles see [Supported Styles](#supported-styles) below.

### Column-Specific Formatters

You can specify a [Formatter](https://developer.apple.com/reference/foundation/nsformatter) on a per-column basis. In this example a [DateFormatter](https://developer.apple.com/reference/foundation/nsdateformatter) is used to customize the display of the `Birthday` column.

```Swift
let df = DateFormatter()
dateFormatter.dateStyle = .medium

let table = TextTable<Person> {
    [Column("Name" <- $0.name),
     Column("Age" <- $0.age),
     Column("Birthday" <- $0.birhtday, formatter: df)]
}

table.print(data, style: Style.psql)
```
```
+-------+-----+--------------+
| Name  | Age | Birthday     |
+-------+-----+--------------+
| Alice | 42  | Aug 13, 2016 |
| Bob   | 22  | Aug 13, 2016 |
| Eve   | 142 | Aug 13, 2016 |
+-------+-----+--------------+
```

### Column Width

By default TextTable will calculate the necessary width for each column. You can also explictely set the width with the `width:` argument.

```Swift
let table = TextTable<Person> {
    [Column("Name" <- $0.name, width: 10),
     Column("Age" <- $0.age, width: 10)]
}
	
table.print(data, style: Style.psql)
```
```
+------------+------------+
| Name       | Age        |
+------------+------------+
| Alice      | 42         |
| Bob        | 22         |
| Eve        | 142        |
+------------+------------+
```

#### Note on Padding 

Some table styles may include padding to the left and/or right of the content, the width argument does not effect include this padding.

### Truncation

#### Truncate Tail

By default, if a width is specified and the contents of the column are wider than the width, the text will be truncated at the tail (`.tail` truncation mode).

```Swift
let table = TextTable<Person> { 
	[Column("Name" <- $0.name, width: 4), // defaults to truncation: .tail
	 Column("Age" <- $0.age)]
}

table.print(data, style: testStyle)
```
```
+------+-----+
| Name | Age |
+------+-----+
| Ali… | 42  |
| Bob  | 22  |
| Eve  | 142 |
+------+-----+
```

#### Truncate Head

You can also truncate at the head of the text string by specifying `.head` for the `truncation:` argument. If no `width` argument is present, the `truncation` argument has no effect.

```Swift
let table = TextTable<Person> {
	[Column("Name" <- $0.name, width: 4, truncate: .head),
	 Column("Age" <- $0.age)]
}

table.print(data, style: testStyle)
```
```
+------+-----+
| Name | Age |
+------+-----+
| …ice | 42  |
| Bob  | 22  |
| Eve  | 142 |
+------+-----+
```

#### Truncate Error

If you prefer to have an fatal error triggered when the contents of a column do not fit the specified width you can specify the `.error` truncation mode.

```Swift
let table = TextTable<Person> {
	[Column("Name" <- $0.name, width: 4, truncate: .error),
	 Column("Age" <- $0.age)]
}

table.print(data, style: testStyle)

// fatal error: Truncation error
```

### Column Alignment

By default, all columns will be left-aligned. You can set the alignment on each column individually.

```Swift
let table = TextTable<Person> {
    [Column("Name" <- $0.name), // default: .left
     Column("Age" <- $0.age, align: .right)]
}

table.print(data, style: Style.psql)
```
```
+-------+-----+
| Name  | Age |
+-------+-----+
| Alice |  42 |
| Bob   |  22 |
| Eve   | 142 |
+-------+-----+
```

## Performance Characteristics

By default TextTable will calculate the necessary width for each column that does not specify an explicit width using the `width:` column argument. This calculation is accomplished by iterating over each row of the table twice: the first time to calculate the max width, the second time to actually render the table. The overhead of this calculation may not be appropriate for very large tables. If you wish to avoid this calculation you must specify the width for every column.

In the example below `table1` will incur the overhead of the width calculation because one of the columns does not have an explicit width. On the other hand, `table2` will not incur the overhead because all columns have an explicit width.

```Swift
// This will still result in extra calculations
let table1 = TextTable<Person> {
    [Column("Name" <- $0.name), 
     Column("Age" <- $0.age, width: 10)]
}
```
```Swift
// This will skip the width calculations
let table2 = TextTable<Person> {
    [Column("Name" <- $0.name, width: 9), 
     Column("Age" <- $0.age, width: 10)]
}
```

## Supported Styles

TextTable supports most of the styles as [tabulate](https://pypi.python.org/pypi/tabulate).

### Plain

```
table.print(data, style: Style.plain)
```
```
Name  Age Birthday
Alice  42  8/13/16
Bob    22  8/13/16
Eve   142  8/13/16
```

### Simple

`simple` is the default Style. It corresponds to `simple_tables` in [Pandoc](http://pandoc.org/) Markdown extensions.

```
table.print(data) // or:
table.print(data, style: Style.simple)
```
```
Name  Age Birthday
----- --- --------
Alice  42  8/13/16
Bob    22  8/13/16
Eve   142  8/13/16
```

### Grid

`grid` follows the conventions of Emacs’ [table.el package](https://www.emacswiki.org/emacs/TableMode). It corresponds to `grid_tables` in Pandoc Markdown extensions.

```
table.print(data, style: Style.grid)
```
```
+-------+-----+----------+
| Name  | Age | Birthday |
+=======+=====+==========+
| Alice |  42 |  8/13/16 |
+-------+-----+----------+
| Bob   |  22 |  8/13/16 |
+-------+-----+----------+
| Eve   | 142 |  8/13/16 |
+-------+-----+----------+
```

### FancyGrid

```
table.print(data, style: Style.fancyGrid)
```
```
╒═══════╤═════╤══════════╕
│ Name  │ Age │ Birthday │
╞═══════╪═════╪══════════╡
│ Alice │  42 │  8/13/16 │
├───────┼─────┼──────────┤
│ Bob   │  22 │  8/13/16 │
├───────┼─────┼──────────┤
│ Eve   │ 142 │  8/13/16 │
╘═══════╧═════╧══════════╛
```

### Psql

```
table.print(data, style: Style.psql)
```
```
+-------+-----+----------+
| Name  | Age | Birthday |
+-------+-----+----------+
| Alice |  42 |  8/13/16 |
| Bob   |  22 |  8/13/16 |
| Eve   | 142 |  8/13/16 |
+-------+-----+----------+
```

### Pipe

`pipe` follows the conventions of [PHP Markdown Extra extension](https://michelf.ca/projects/php-markdown/extra/). It corresponds to `pipe_tables` in Pandoc. This style uses colons to indicate column alignment.

```
table.print(data, style: Style.pipe)
```
```
| Name  | Age | Birthday |
|:------|----:|---------:|
| Alice |  42 |  8/13/16 |
| Bob   |  22 |  8/13/16 |
| Eve   | 142 |  8/13/16 |
```

### Org

`org` follows the conventions of [Emacs org-mode](http://orgmode.org/).

```
table.print(data, style: Style.org)
```
```
| Name  | Age | Birthday |
|-------+-----+----------|
| Alice |  42 |  8/13/16 |
| Bob   |  22 |  8/13/16 |
| Eve   | 142 |  8/13/16 |
```

### Rst

`rst` follows the conventions of simple table of the [reStructuredText](http://docutils.sourceforge.net/rst.html) Style.

```Swift
table.print(data, style: Style.rst)
```
```
===== === ========
Name  Age Birthday
===== === ========
Alice  42  8/13/16
Bob    22  8/13/16
Eve   142  8/13/16
===== === ========
```

### Html

`html` produces [HTML table](https://www.w3.org/TR/html5/tabular-data.html) markup.

```
table.print(data, style: Style.html)
```
```
<table>
    <tr>
        <th style="text-align:left;">Name</th>
        <th style="text-align:center;">Age</th>
        <th style="text-align:right;">Birthday</th>
    </tr>
    <tr>
        <td>Alice</td>
        <td>42</td>
        <td>8/14/16</td>
    </tr>
    <tr>
        <td>Bob</td>
        <td>22</td>
        <td>8/14/16</td>
    </tr>
    <tr>
        <td>Eve</td>
        <td>142</td>
        <td>8/14/16</td>
    </tr>
</table>
```

### Latex

`latex` produces a [tabular](https://www.tug.org/TUGboat/tb28-3/tb90hoeppner.pdf) environment for [Latex](https://www.latex-project.org/about/).

```
table.print(data, style: Style.latex)
```
```
\begin{tabular}{lll}
\hline
 Name & Age & Birthday \\
\hline
 Alice & 42 & 8/13/16 \\
 Bob & 22 & 8/13/16 \\
 Eve & 142 & 8/13/16 \\
\hline
\end{tabular}
```

## Custom Styles

You can create a custom table style by conforming to the `TextTableStyle` protocol.


```Swift
enum MyCustomStyle: TextTableStyle {
    // implement methods from TextTableStyle...
}

extension Style {
    static let custom = MyCustomStyle.self
}

table.print(data, style: Style.custom)
```

## Changes

A list of changes can be found in the [CHANGELOG](CHANGELOG.md).

## Alternatives

Other Swift libraries that serve a similar purpose.

* [SwiftyTextTable](https://github.com/scottrhoyt/SwiftyTextTable)

## License

MIT License

Copyright (c) 2016 Cristian Filipov

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


CHANGELOG.md
# Change Log

## [1.0.0-alpha.4](https://github.com/cfilipov/TextTable/releases/tag/v1.0.0-alpha.4)

### What's New

* Update for Xcode 8 beta 6 (Apple Swift version 3.0 (swiftlang-800.0.43.6 clang-800.0.38))
* Travis-CI integration

### SPM Dependency Snippet

```Swift
.Package(url: "https://github.com/cfilipov/TextTable", Version(1, 0, 0, prereleaseIdentifiers: ["alpha", "4"]))
```

## [1.0.0-alpha.3](https://github.com/cfilipov/TextTable/releases/tag/v1.0.0-alpha.3)

### What's New

* **This release contains breaking changes to the API**. This should hopefully be the last major breaking change to the API for a long time.
* `TextTable<T>` construction closure now takes an instance of `T` instead of a `Config<T>`. `Config<T>` has been removed completely. Instead of calling `column(...)` on a `Config<T>` in the closure, you now return an array of `Column` instances. See example transition below for details.
* **Fixed**: Missing/broken tests.
* **Fixed**: Latex `tabular` environment contains incorrect alignments.
* **Fixed**: Handle optional values in mapping.
* **Fixed**: `FancyGrid` using pipe symbol instead of unicode drawing character for header separator.
* All columns are now left-aligned by default (instead of right-aligning the first column by default).
* `TextTableFormatter` has been renamed to `TextTableStyle` to avoid confusion with the Foundation [Formatters](https://developer.apple.com/reference/foundation/nsformatter) class. All references to the word "format" when referring to the type of table rendered have bee changed to "style".
* The `TextTableStyle` protocol (formerly `TextTableFormatter `) is now simplified. All methods in `TextTableStyle` are now static, no state is maintained in `TextTableStyle`. The methods have also been consolidated (`row` instead of `beginRow`, `endRow`, `content`, `beginColumn`, etc...).
* `TextTable<T>` and all the internal types are now value types instead of classes.
* All built-in styles are case-less `enums` so they cannot be accidentally instantiated.

### Known Issues

* Very little effort has but put into performance optimizations. String utilities in particular.
* It should be possible to create columns without headers, but this hasn't been tested and likely doesn't work yet.

### Transitioning

Before:

```Swift
let table = TextTable<Person> { t in
    t.column("Name") { $0.name }
    t.column("Age") { $0.age }
    t.column("Birthday") { $0.birhtday }
}
```

After:

```Swift
let table = TextTable<Person> {
    [Column("Name" <- $0.name),
     Column("Age" <- $0.age),
     Column("Birthday" <- $0.birhtday)]
}
```

### SPM Dependency Snippet

```Swift
.Package(url: "https://github.com/cfilipov/TextTable", Version(1, 0, 0, prereleaseIdentifiers: ["alpha", "3"]))
```

## [1.0.0-alpha.2](https://github.com/cfilipov/TextTable/releases/tag/v1.0.0-alpha.2)

### What's New

* Column truncation support. There is now an optional `truncate:` argument to `width`.
* Added some documentation.

### Known Issues

* Very little effort has but put into performance optimizations.
* It should be possible to create columns without headers, but this hasn't been tested and likely doesn't work yet.

### SPM Dependency Snippet

```Swift
.Package(url: "https://github.com/cfilipov/TextTable", Version(1, 0, 0, prereleaseIdentifiers: ["alpha", "2"]))
```

## [1.0.0-alpha.1](https://github.com/cfilipov/TextTable/releases/tag/v1.0.0-alpha.1)

### What's New

* Support for center alignment.
* If all columns have explicit width, then width calculations are skipped.
* Escape strings in certain formats (HTML & Latex, for example).
* Fixed: `TextTable` config persists calculated column widths. This results in widths getting stuck from the first call to `print` or `string(for:)`.

### Known Issues

* Very little effort has but put into performance optimizations.
* It should be possible to create columns without headers, but this hasn't been tested and likely doesn't work yet.

### SPM Dependency Snippet

```Swift
.Package(url: "https://github.com/cfilipov/TextTable", Version(1, 0, 0, prereleaseIdentifiers: ["alpha", "1"]))
```

## [1.0.0-alpha.0](https://github.com/cfilipov/TextTable/releases/tag/v1.0.0-alpha.0)

### What's New

* Breaking change: completely re-written API. No more conforming to a protocol, instead a `TextTable` class is used in a similar way to `NSFormatter`. This offers much more flexibility.
* Many more output formats. Similar to what you get from Python's [tabulate](https://pypi.python.org/pypi/tabulate) lib.

### Known Issues

* No attempt at optimizing performance has been made yet. Even when widths are provided, an expensive calculation is still performed.
* No escaping is being done. None of the formatters even attempt to sanitize the input strings.
* Center alignment not supported. This will result in a `fatalError()`.
* It should be possible to create columns without headers, but this hasn't been tested and likely doesn't work yet.

### SPM Dependency Snippet

```Swift
.Package(url: "https://github.com/cfilipov/TextTable", Version(1, 0, 0, prereleaseIdentifiers: ["alpha", "0"]))
```

## [0.0.0](https://github.com/cfilipov/TextTable/releases/tag/v0.0.0)

Initial release. 

Sources/TextTable.swift
//
//  TextTablePrintable.swift
//  Stooli
//
//  Created by Cristian Filipov on 8/8/16.
//
//

import Foundation

infix operator <- : AssignmentPrecedence

public func <- (left: String, right: Any) -> (title: String, value: Any) {
    return (left, right)
}

private func + (left: [Column], right: [Int]) -> [Column] {
    precondition(left.count == right.count)
    return zip(left, right).map { $0.0.settingWidth($0.1) }
}

private func unwrap(_ any: Any) -> Any {
    let m = Mirror(reflecting: any)
    if m.displayStyle != .optional { return any }
    if m.children.count == 0 { return "NULL" }
    let (_, some) = m.children.first!
    return some
}

public enum Alignment {
    case left
    case right
    case center
}

public enum Truncation {
    case tail
    case head
    case error
}

/**
 The style pseudo-namespace groups together all the built-in styles that come siwht TextTable. You can add you own styles by extending the `Style` type.
 */
public enum Style {
    public static let simple = Simple.self
    public static let plain = Plain.self
    public static let rst = Rst.self
    public static let psql = Psql.self
    public static let pipe = Pipe.self
    public static let org = Org.self
    public static let latex = Latex.self
    public static let html = Html.self
    public static let grid = Grid.self
    public static let fancy = FancyGrid.self
}

/**
 A `Column` defines how a particular piece of data will be rendered in the text table.
 */
public struct Column {
    let title: String
    let value: Any
    let width: Int?
    let align: Alignment
    let truncate: Truncation
    let formatter: Formatter?

    public init(title: String, value: Any, width: Int? = nil, align: Alignment = .left, truncate: Truncation = .tail, formatter: Formatter? = nil) {
        self.title = title
        self.value = unwrap(value)
        self.width = width
        self.align = align
        self.truncate = truncate
        self.formatter = formatter
    }

    public init(_ mapping: @autoclosure () -> (String, Any), width: Int? = nil, align: Alignment = .left, truncate: Truncation = .tail, formatter: Formatter? = nil) {
        let (t, v) = mapping()
        self = Column(
            title: t,
            value: v,
            width: width,
            align: align,
            truncate: truncate,
            formatter: formatter)
    }

    internal var resolvedWidth: Int {
        return width ?? title.characters.count
    }

//    private func string(for value: Any) -> String {
//        var string = String(describing: value)
//        if let formatter = formatter {
//            string = formatter.string(for: value) ?? string
//        }
//        return string
//    }

    fileprivate func settingWidth(_ newWidth: Int) -> Column {
        return Column(
            title: self.title,
            value: self.value,
            width: newWidth,
            align: self.align,
            truncate: self.truncate,
            formatter: self.formatter)
    }
}

public extension Column {
    func string(for style: TextTableStyle.Type) -> String {
        var string = ""
        if let formatter = formatter {
            string = formatter.string(for: value) ?? string
        } else {
            string = String(describing: value)
        }
        return style.prepare(string, for: self)
    }

    func headerString(for style: TextTableStyle.Type) -> String {
        return style.prepare(title, for: self)
    }

    func repeated(_ string: String) -> String {
        return String(repeating: string, count: resolvedWidth)
    }
}

/**
 The `TextTableStyle` protocol defines a set of static methods used to render a textual table. Styles are meant to be stateless and re-usable and are not meant to be instantiated. For this reason it is recomended to use case-less enums for your custom styles.
 
 Each static method accepts an inout `String` argument which is to be appended to as the output and an array of columns.
 */
public protocol TextTableStyle {
    /**
     Prepare the string to be rendered. The string may be the header text or column value. Typically this method will pad, truncate and call `escape()` on the string. If no special preperation is required for this style, you can simply return the passed-in string.
     */
    static func prepare(_ s: String, for column: Column) -> String

    /**
     Escape any special characters from the provided string. Most styles simply return the passed-in string. However, in some cases you might want to escape certain characters that have special meaning in your output format. For example, the HTML style will escape `<`, `>` and other reserved HTML characters.
     */
    static func escape(_ table: String) -> String

    /**
     Called before the header or any of the rows.
     */
    static func begin(_ table: inout String, index: Int, columns: [Column])

    /**
     Called after all the rows have completed.
     */
    static func end(_ table: inout String, index: Int, columns: [Column])

    /**
     Called just before the first row.
     */
    static func header(_ table: inout String, index: Int, columns: [Column])

    /**
     Called for each row.
     */
    static func row(_ table: inout String, index: Int, columns: [Column])
}

/**
 `TextTable` formats a string into a textual representation of a table. `TextTable` can render tables in various different styles including HTML, Markdown and Latex tables.

 Instances of `TextTable` represent the mapping of a type to a tabular representation. That is, `TextTable` defines how an instance of `T` can be transformed into a set of columns.
 
 A `TextTable` is configured by returning an array of `Column` instances in the closure argument of the initializer. Each `Column` may specify an width, alignment, truncation mode and NSFormatter. By default, widths are calculated based on the longest content and alignment defaults to left.
 
 ## Basic Usage
 
     struct Person {
         let name: String
         let age: Int
         let birhtday: Date
     }

     let df = DateFormatter()
     dateFormatter.dateStyle = .short

     let table = TextTable<Person> {
         [Column("Name" <- $0.name),
          Column("Age" <- $0.age, width: 6, align: .center),
          Column("Birthday" <- $0.birhtday, formatter: df)]
     }
 
     table.print(data)

 Output:

     Name   Age   Birthday
     ----- ------ --------
     Alice   42    8/13/16
     Bob     22    8/13/16
     Eve    142    8/13/16
 */
public struct TextTable<T> {
    /**
     The adapter function is responsible for mapping an instance `T` to an array of columns. This will be called for each row, including the header. The adapter function may actually be called multiple times for each row or header. You should not make any assumptions regarding the context, order or number of calls.
     */
    public typealias Adapter = (T) -> [Column]

    private let adapter: Adapter

    /**
     Creates an instance of `TextTable` which is used to format strings of tables. This instance represents the mapping of a type `T` to its corresponding columns. An instance can be re-used (and should be re-used unless you need to change the mapping or column configuration).
     */
    public init(_ adapter: Adapter) {
        self.adapter = adapter
    }

    private func calculateWidths<C: Collection>(for data: C, style: TextTableStyle.Type) -> [Int] where C.Iterator.Element == T {
        guard let first = data.first else { return [] }
        let headerCols = adapter(first)
        var widths = headerCols.map{$0.width ?? 0}
        for (index,column) in headerCols.enumerated() {
            if let w = column.width {
                widths[index] = w
            } else {
                let text = column.headerString(for: style)
                widths[index] = max(text.characters.count, widths[index])
            }
        }
        for element in data {
            let cols = adapter(element)
            for (index, column) in cols.enumerated() {
                if let w = column.width {
                    widths[index] = w
                } else {
                    let text = column.string(for: style)
                    widths[index] = max(text.characters.count, widths[index])
                }
            }
        }
        return widths
    }

    /**
     Returns a string representing the data rendered as a textual table.
     
     - parameter data: A collection of elements `T` for which this `TextTable` instance has been configured.
     - parameter style: The style of table to be rendered. See `Style` for more options.
     */
    public func string<C: Collection>(for data: C, style: TextTableStyle.Type = Style.simple) -> String? where C.Iterator.Element == T {
        guard let first = data.first else { return nil }
        var table = ""
        let cols = adapter(first)
        var widths = cols.flatMap{$0.width}
        if widths.count < cols.count {
            widths = calculateWidths(for: data, style: style)
        }
        style.begin(&table, index: -1, columns: cols + widths)
        style.header(&table, index: -1, columns: cols + widths)
        for (index, element) in data.enumerated() {
            style.row(&table, index: index, columns: adapter(element) + widths)
        }
        style.end(&table, index: -1, columns: cols + widths)
        return table
    }

    /**
     Prints the data rendered as a textual table.
     
     - parameter data: A collection of elements `T` for which this `TextTable` instance has been configured.
     - parameter style: The style of table to be rendered. See `Style` for more options.
     */
    public func print<C: Collection>(_ data: C, style: TextTableStyle.Type = Style.simple) where C.Iterator.Element == T {
        let table = string(for: data, style: style)!
        Swift.print(table)
    }
}
